File: Check_sum.java
import java.util.Scanner;

public class Check_sum {

    public static byte calculateChecksum(String data) {
        int sum = 0;
        for (int i = 0; i < data.length(); i++) {
            sum += (byte) data.charAt(i);
        }
        return (byte) (sum % 256);
    }

    public static byte[] sendFrame(String data) {
        byte checksum = calculateChecksum(data);
        byte SOF = 0x7E;
        byte EOF = 0x7F;

        byte[] dataBytes = data.getBytes();
        byte[] frame = new byte[dataBytes.length + 3];

        frame[0] = SOF;
        System.arraycopy(dataBytes, 0, frame, 1, dataBytes.length);
        frame[1 + dataBytes.length] = checksum;
        frame[2 + dataBytes.length] = EOF;

        return frame;
    }

    public static boolean receiveFrame(byte[] frame, StringBuilder outData) {
        byte SOF = 0x7E;
        byte EOF = 0x7F;

        if (frame[0] != SOF || frame[frame.length - 1] != EOF) {
            System.out.println("Frame error: Invalid framing bytes.");
            return false;
        }

        int dataLen = frame.length - 3;
        byte[] dataBytes = new byte[dataLen];
        System.arraycopy(frame, 1, dataBytes, 0, dataLen);

        String data = new String(dataBytes);
        outData.append(data);

        byte receivedChecksum = frame[1 + dataLen];
        byte calcChecksum = calculateChecksum(data);

        if (receivedChecksum != calcChecksum) {
            System.out.println("Checksum error!");
            return false;
        }

        return true;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter data to send: ");
        String data = sc.nextLine();

        byte[] frame = sendFrame(data);

        System.out.print("Transmitted Frame (in hex): ");
        for (byte b : frame) {
            System.out.printf("%02X ", b);
        }
        System.out.println();

        StringBuilder receivedData = new StringBuilder();
        if (receiveFrame(frame, receivedData)) {
            System.out.println("Received data: " + receivedData);
            System.out.println("No error detected in frame.");
        }

        sc.close();
    }
}

File: Check_sum.py
from typing import Optional
def calculate_checksum(data: str) -> int:
    """Calculate checksum as sum of bytes modulo 256"""
    checksum = sum(bytearray(data, "utf-8")) % 256
    return checksum

def send_frame(data: str) -> bytes:
    """Create frame with SOF, data, checksum, and EOF"""
    SOF = 0x7E
    EOF = 0x7F
    checksum = calculate_checksum(data)

    data_bytes = data.encode("utf-8")
    frame = bytearray()
    frame.append(SOF)
    frame.extend(data_bytes)
    frame.append(checksum)
    frame.append(EOF)

    return bytes(frame)

def receive_frame(frame: bytes) -> Optional[str]:
    """Check framing and checksum at receiver side"""
    SOF = 0x7E
    EOF = 0x7F

    if frame[0] != SOF or frame[-1] != EOF:
        print("Frame error: Invalid framing bytes.")
        return None

    data_bytes = frame[1:-2]
    received_checksum = frame[-2]
    data = data_bytes.decode("utf-8")

    calc_checksum = calculate_checksum(data)

    if received_checksum != calc_checksum:
        print("Checksum error!")
        return None

    return data

def main():
    data = input("Enter data to send: ")

    frame = send_frame(data)
    print("Transmitted Frame (in hex):", " ".join(f"{b:02X}" for b in frame))

    received_data = receive_frame(frame)
    if received_data is not None:
        print("Received data:", received_data)
        print("No error detected in frame.")

if __name__ == "__main__":
    main()
